core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -93,59 +90,61 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) throws Ex
-                                                                                    endpointWithPort.port()));
-            if (resolveFuture.isDone()) {
-                finishResolve(ctx, endpointWithPort, resolveFuture, req, res, timingsBuilder);
-            } else {
-                resolveFuture.addListener(
-                        (FutureListener<InetSocketAddress>) future ->
-                                finishResolve(ctx, endpointWithPort, future, req, res, timingsBuilder));
-            }
+            resolveAddress(endpointWithPort, ctx).handle((resolved, cause) -> {
+                timingsBuilder.dnsResolutionEnd();
+                if (cause == null) {
+                    acquireConnectionAndExecute(ctx, resolved, req, res, timingsBuilder);
+                } else {
+                    ctx.logBuilder().session(null, ctx.sessionProtocol(), timingsBuilder.build());
+                    final UnprocessedRequestException wrappedCause = UnprocessedRequestException.of(cause);
+                    handleEarlyRequestException(ctx, req, wrappedCause);
+                    res.close(wrappedCause);
+                }
+                return null;
+            });
         }
 
         return res;
     }
 
-    private void finishResolve(ClientRequestContext ctx, Endpoint endpointWithPort,
-                               Future<InetSocketAddress> resolveFuture, HttpRequest req,
-                               DecodedHttpResponse res, ClientConnectionTimingsBuilder timingsBuilder) {
-        timingsBuilder.dnsResolutionEnd();
+    private CompletableFuture<Endpoint> resolveAddress(Endpoint endpoint, ClientRequestContext ctx) {
+        // IP address has not been resolved yet.
+        assert !endpoint.hasIpAddr() && endpoint.hasPort();
+
+        final Future<InetSocketAddress> resolveFuture =
+                addressResolverGroup.getResolver(ctx.eventLoop().withoutContext())
+                                    .resolve(InetSocketAddress.createUnresolved(endpoint.host(),
+                                                                                endpoint.port()));
         if (resolveFuture.isSuccess()) {
-            final String ipAddr = resolveFuture.getNow().getAddress().getHostAddress();
-            acquireConnectionAndExecute(ctx, endpointWithPort, ipAddr, req, res, timingsBuilder);
+            final InetAddress address = resolveFuture.getNow().getAddress();
+            return CompletableFuture.completedFuture(endpoint.withIpAddr(address));
         } else {
-            ctx.logBuilder().session(null, ctx.sessionProtocol(), timingsBuilder.build());
-            final UnprocessedRequestException cause = UnprocessedRequestException.of(resolveFuture.cause());
-            handleEarlyRequestException(ctx, req, cause);
-            res.close(cause);
+            final CompletableFuture<Endpoint> endpointFuture = new CompletableFuture<>();
```

Comment:
Can't we just use `Future` instead of creating a new `CompletableFuture`? We might create a new endpoint in the listener of the `Future`.

Reply:
This method returns a new `Endpoint` with a resolved IP address. So, anyway, it needs to return a new `(Completable)Future` or take a callback.
Let me use a callback to avoid creating an additional future.